### ISSUE TITLE
listen failure

### ISSUE THREAD

**Issue by chrishough** [guard/guard#811]
No matter how I adjust my guard file I can not seem to fix this... 
```
[2016-01-11 16:32:16] INFO  WEBrick 1.3.1
[2016-01-11 16:32:16] INFO  ruby 2.3.0 (2015-12-25) [x86_64-darwin15]
[2016-01-11 16:32:16] INFO  WEBrick::HTTPServer#start: pid=93168 port=3000
16:32:19 - INFO - Rails started, pid 93168

16:32:19 - INFO - Guard is now watching at '/Users/chrishough/Studio/Code-Consulting/bilingualhire'
[1] Backend guard(main)>         ** ERROR: directory is already being watched! **

        Directory: /Users/chrishough/Studio/Code-Consulting/bilingualhire/.bundle/ruby/2.2.0/gems/libv8-3.16.14.13/vendor/v8/build/gyp

        is already being watched through: /Users/chrishough/Studio/Code-Consulting/bilingualhire/.bundle/ruby/2.2.0/gems/libv8-3.16.14.13/vendor/gyp

        MORE INFO: https://github.com/guard/listen/wiki/Duplicate-directory-errors
        ** ERROR: directory is already being watched! **

        Directory: /Users/chrishough/Studio/Code-Consulting/bilingualhire/.bundle/ruby/2.3.0/gems/libv8-3.16.14.13/vendor/v8/build/gyp

        is already being watched through: /Users/chrishough/Studio/Code-Consulting/bilingualhire/.bundle/ruby/2.3.0/gems/libv8-3.16.14.13/vendor/gyp

        MORE INFO: https://github.com/guard/listen/wiki/Duplicate-directory-errors
```
Has anyone encountered this? I have tried the fixes listed but nothing works...

**Comment by e2** [comment]
Yes, it's a known problem.

Quick fix: remove the .bundle directory and install gems globally with `bundle install`.

You don't want copies of gems in your project.


Here's what's going on: libv8 gem uses a symlink, which causes it's directories to be listed twice, but this isn't recommended in Listen. Sadly, due to a poor file monitoring implementation on OSX, Listen can't ignore such duplicated directories, and this leads to many terrible problems for users.

Ideally, you don't want have the `.bundle` directory during development. And that's the best solution here.

**Comment by chrishough** [comment]
Hmm... when using rbenv I thought it was a preferred practice to install gems locally to the project.  Am I missing something here?

**Comment by e2** [comment]
On OSX, all folders are watched recusively. That's the implementation restriction. 

You can instead use the `directories` statements to watch all the other folders except the `.bundle` directory, but you'd also have to symlink any files in the project root directory that you'd like to watch (like `Guardfile`, etc.).

I you must have the `.bundler` directory (I don't know about rbenv), I'd suggest manually replacing that symlink in libv8 into a copy of the linked files.

Another idea is to use polling mode in Guard - but that can be slow with all the gems being scanned every fraction of a second. Not to mention CPU usage on OSX.

Overall, watching tens of thousands of files you don't care about will just slow up startup time and response time while you're working - with no upside.

Personally I use RVM. The biggest issue with is was the PATH statement screwed up by installer of oh-my-zsh. But with that fixed (and setting a login shell), I haven't complained about it.

**Comment by chrishough** [comment]
Do you have an example setup you are referencing? I tried ignoring `.bundle` and it did not work.

**Comment by e2** [comment]
You can't ignore it on OSX. 

Check the default Guardfile template for the `directories` clause. If that's what you mean.

Here are your options (in order of what I'd recommend):

1. remove `.bundle` directory completely and use global gems (recommended)
2. reconfigure `.bundle` dir outside your project (see: http://bundler.io/man/bundle-config.1.html)
3. copy files to avoid the symlink - hacky, but you'll avoid the problem quickly (until you reinstall the gem, etc.)
4. use `directories` in your `Guardfile` (see guard docs and wiki) to watch just selected dirs, like `app`, `lib`, `spec`, etc.
5. use polling (slow)
6. use Linux ? (my solution)
7. symlinking patches for Listen (I don't intend to implement this myself)
8. ?

I just don't recommend having a `.bundle` directory and watching it on OSX.

Long story short, OSX is the problem here. Everything else is a workaround for it.

**Comment by chrishough** [comment]
hmmm... I am not sure that approach will work for rbenv... in fact I could see that blowing up projects.. 

I am already using directories but it does not seem to be respecting them... 

```ruby
ignore %r{^ignored/path/}, /public/
filter /\.txt$/, /.*\.zip/

guard 'livereload' do
  extensions = {
    css: :css,
    scss: :css,
    sass: :css,
    js: :js,
    coffee: :js,
    html: :html,
    png: :png,
    gif: :gif,
    jpg: :jpg,
    jpeg: :jpeg,
  }
  rails_view_exts = %w(erb haml slim)
  # file types LiveReload may optimize refresh for
  compiled_exts = extensions.values.uniq
  watch(%r{public/.+\.(#{compiled_exts * '|'})})
  extensions.each do |ext, type|
    watch(%r{
          (?:app)
          (?:/assets/\w+/(?<path>[^.]+) # path+base without extension
           (?<ext>\.#{ext})) # matching extension (must be first encountered)
          (?:\.\w+|$) # other extensions
          }x) do |m|
      path = m[1]
      "/assets/#{path}.#{type}"
    end
  end
  # file needing a full reload of the page anyway
  watch(%r{app/views/.+\.(#{rails_view_exts * '|'})$})
  watch(%r{app/helpers/.+\.rb})
  watch(%r{config/locales/.+\.yml})
end

# guard 'rake', task: 'bilingualhire:test:jshint' do
#   watch(%r{(app|vendor)(/assets/\w+/(.+\.(css|js|html))).*}) { |m| "/assets/#{m[3]}" }
# end
#
# guard 'rake', task: 'bilingualhire:test:scsslint' do
#   watch(%r{(app|vendor)(/assets/\w+/.+)\.s(c|a)ss}) { |m| "#{m[2]}.css" }
# end
```

**Comment by e2** [comment]
By `directories` I mean uncommenting this: https://github.com/guard/guard/blob/master/lib/guard/templates/Guardfile#L5

**Comment by chrishough** [comment]
Ahh, that worked. I just set the directories I wanted to watch.  So far so good. Thanks @e2

**Comment by asecondwill** [comment]
if the problem here is OSX, how come it wasn't a problem until recently?

**Comment by e2** [comment]
@asecondwill - it's better to have an error (now) than have insanely high CPU usage on big projects on OSX.

Do you have a specific problem that you're getting this?
